File: ListFiles.java
/*

		Copyright 2014 Krrishnaaaa
		
	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at
	http://www.apache.org/licenses/LICENSE-2.0
	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.InputStream;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

public class ListFiles {

	public static void main(String[] args) {
		if(args.length == 0) {
			System.out.println("Provide directory to search");
			return;
		}
		ArrayList<Files> mFiles = new ArrayList<Files>();

		String rootPath = args[0];

		File folder = new File(rootPath);

		System.out.println("\nSearching for duplicate files on path: " + rootPath);

		System.out.println("Processing started at " + new Date());

		addFileToList(folder, mFiles);

		ArrayList<DuplicateDetails> dups = new ArrayList<DuplicateDetails>();

		for (int i = 0; i < mFiles.size(); i++) {
			Set<String> completePath = new HashSet<String>();
			DuplicateDetails details = new DuplicateDetails();
			for (int j = i + 1; j < mFiles.size(); j++) {
				Files f1 = mFiles.get(i);
				Files f2 = mFiles.get(j);
				if (f1.equals(f2)) {
					completePath.add(f1.getFilePath());
					completePath.add(f2.getFilePath());
					mFiles.remove(j);
					j--;
				}
				if (completePath.size() > 0) {
					details.filePath = completePath;
					details.md5 = f1.getMd5();
				}
			}
			if (completePath.size() > 0) {
				dups.add(details);
			}
		}

		System.out.println("\nDuplicate(s) of (" + dups.size() + ") file(s) found.");
		String fileName = "/home/krishna/toDelete"
							 + args[0].replaceAll(File.separator, "_").replaceAll(" ", "_");
		System.out.println("\nList of duplicate files is stored in " + fileName);
		File logFile = new File(fileName);
		BufferedWriter writer = null;
		try {
			writer = new BufferedWriter(new FileWriter(logFile));

			for (DuplicateDetails details : dups) {
				writer.write(details.md5+"\n");
				for (String completePath : details.filePath) {
					writer.write(completePath+"\n");
				}
			}
		} catch (Exception e) {
		    e.printStackTrace();
		} finally {
		    try {
		        if(writer != null) writer.close();
		    } catch (Exception e) {
		    }
		}
		
		System.out.println("\nProcessing ended at " + new Date());
	}
	
	private static int counter = 0;
	
	private static void addFileToList(File mFile, ArrayList<Files> mFiles) {
		if (mFile.isFile()) {
			counter++;
			String path = mFile.getAbsolutePath();
			double kb = mFile.length()/1024.0;
			String message = String.format("\r(%d) files scanned. Current file size : %10.2f KB", counter, kb);
			System.out.print(message);
			addFileToList(path, mFiles);
		} else if (mFile.isDirectory()) {
			for (File child : mFile.listFiles()) {
				addFileToList(child, mFiles);
			}
		}
	}
	
	static void addFileToList(String filePath, ArrayList<Files> mFiles) {
		String md5 = fileToMD5(filePath);
		Files files = new Files();
		files.setFilePath(filePath);
		files.setMd5(md5);
		mFiles.add(files);
	}

	public static String fileToMD5(String filePath) {
		InputStream inputStream = null;
		try {
			inputStream = new FileInputStream(filePath);
			byte[] buffer = new byte[1024];
			MessageDigest digest = MessageDigest.getInstance("MD5");
			int numRead = 0;
			while (numRead != -1) {
				numRead = inputStream.read(buffer);
				if (numRead > 0)
					digest.update(buffer, 0, numRead);
			}
			byte[] md5Bytes = digest.digest();
			return convertHashToString(md5Bytes);
		} catch (Exception e) {
			return null;
		} finally {
			if (inputStream != null) {
				try {
					inputStream.close();
				} catch (Exception e) {
				}
			}
		}
	}

	private static String convertHashToString(byte[] md5Bytes) {
		String returnVal = "";
		for (int i = 0; i < md5Bytes.length; i++) {
			returnVal += Integer.toString((md5Bytes[i] & 0xff) + 0x100, 16).substring(1);
		}
		return returnVal.toUpperCase();
	}

}

File: src/main/kotlin/com/pcsalt/utility/ListFiles.kt
/*

		Copyright 2014 Krrishnaaaa

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at
	http://www.apache.org/licenses/LICENSE-2.0
	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

package com.pcsalt.utility

import java.io.BufferedWriter
import java.io.File
import java.io.FileInputStream
import java.io.FileWriter
import java.security.MessageDigest
import java.util.Date
import java.util.Locale

data class Files(val md5: String, val filePath: String)
data class DuplicateDetails(val md5: String, val filePath: Set<String>)

class ListFiles {
  private var counter = 0
  private var totalFilesCount = 0
  private var lastMessageLength = 0

  fun find(dirToSearch: List<String>) {

    val mFiles = mutableListOf<Files>()
    val duplicates = mutableMapOf<String, MutableSet<String>>() // Map for grouping duplicates
    val ignoreFolders = setOf(".git", "build", "node_modules", ".gradle", ".idea") // Specify folders to ignore
    val ignoreFiles = setOf(".DS_Store")

    println("\nSearching for duplicate files in directories: ${dirToSearch.joinToString(", ")}")
    println("Processing started at " + Date())
    for (rootPath in dirToSearch) {
      countFiles(File(rootPath), ignoreFolders, ignoreFiles)
    }

    println("Total files to scan: $totalFilesCount")
    // Process each directory
    for (rootPath in dirToSearch) {
      addFilesToList(File(rootPath), mFiles, ignoreFolders, ignoreFiles)
    }

    // Group files by MD5 hash
    for (file in mFiles) {
      duplicates.computeIfAbsent(file.md5) { mutableSetOf() }.add(file.filePath)
    }

    // Filter out non-duplicates
    val duplicateDetails = duplicates.filter { it.value.size > 1 }
      .map { DuplicateDetails(it.key, it.value) }

    println("\nDuplicate(s) of (${duplicateDetails.size}) file(s) found.")
    val logFileName =
      "./toDelete" + dirToSearch.joinToString("_") { it.replace(File.separator.toRegex(), "_").replace(" ", "_") }
    println("\nList of duplicate files is stored in $logFileName")

    // Write duplicates to log file
    writeDuplicatesToFile(logFileName, duplicateDetails)

    println("\nProcessing ended at " + Date())
  }

  private fun countFiles(folder: File, ignoreFolders: Set<String>, ignoreFiles: Set<String>) {
    if (folder.isDirectory) {
      if (ignoreFolders.any { folder.resolve(it).exists() }) {
        return
      }
      folder.listFiles()?.forEach { child -> countFiles(child, ignoreFolders, ignoreFiles) }
    } else if (folder.isFile) {
      if (!ignoreFiles.contains(folder.name)) {
        totalFilesCount++
      }
    }
  }

  private fun addFilesToList(
    folder: File,
    mFiles: MutableList<Files>,
    ignoreFolders: Set<String>,
    ignoreFiles: Set<String>
  ) {
    if (folder.isFile) {
      if (!ignoreFiles.contains(folder.name)) {
        counter++
        val path = folder.absolutePath
        val kb = folder.length() / 1024.0
        if (lastMessageLength > 0) print("\r"+" ".repeat(lastMessageLength))
        val message = String.format("\r(%d) files scanned. %s size : %10.2f KB", counter, folder.path, kb)
        lastMessageLength = message.length
        print(message)
        addFileToList(path, mFiles)
      }
    } else if (folder.isDirectory) {
      if (ignoreFolders.any { folder.resolve(it).exists() }) {
        return
      }
      folder.listFiles()?.forEach { child -> addFilesToList(child, mFiles, ignoreFolders, ignoreFiles) }
    }
  }

  private fun addFileToList(filePath: String, mFiles: MutableList<Files>) {
    val md5 = fileToMD5(filePath)
    mFiles.add(Files(md5, filePath))
  }

  private fun fileToMD5(filePath: String): String {
    FileInputStream(filePath).use { inputStream ->
      val buffer = ByteArray(1024)
      val digest = MessageDigest.getInstance("MD5")
      var numRead: Int
      while (inputStream.read(buffer).also { numRead = it } != -1) {
        if (numRead > 0) digest.update(buffer, 0, numRead)
      }
      val md5Bytes = digest.digest()
      return convertHashToString(md5Bytes)
    }
  }

  private fun convertHashToString(md5Bytes: ByteArray): String {
    val returnVal = StringBuilder()
    for (md5Byte in md5Bytes) {
      returnVal.append(((md5Byte.toInt() and 0xff) + 0x100).toString(16).substring(1))
    }
    return returnVal.toString().uppercase(Locale.getDefault())
  }

  private fun writeDuplicatesToFile(fileName: String, duplicateDetails: List<DuplicateDetails>) {
    val logFile = File(fileName)
    try {
      BufferedWriter(FileWriter(logFile)).use { writer ->
        for (details in duplicateDetails) {
          writer.write("${details.md5}\n")
          for (completePath in details.filePath) {
            writer.write("$completePath\n")
          }
        }
      }
    } catch (e: Exception) {
      e.printStackTrace()
    }
  }
}
